Disallow users from configuring analyzer version

mariha opened the pull request:
I am looking for a general feedback if it's good direction to address issue #18267. 

If user tries to set `index.analysis.version` or `index.analysis.analyzer.*.version` the request fails with an error response:
```
"type": "illegal_argument_exception",
"reason": "illegal settings entry for [index.analysis.version]. Since version 8.0.0 the property is reserved for internal usage only."
```
earlier it was simply stored without any validation prior to usage.
For backwards compatibility I thought of printing a deprecation warning in earlier versions:
```
Setting value for [index.analysis.version] is deprecated and starting from version 8.0.0 will not be allowed any more.
```

Similar validation as for _analyzer_ component could be done for _tokenizer_ and _filters_. Are all settings for analyzers described anywhere?

As far as I understand the flow, with the changes, in addition to the above validation, when trying to use already saved settings with specified [version] it is checked if it is one of Lucene supported versions. If not, warning is logged:
```
Unsupported index version 2.0, using earliest still supported instead -> 7.0.0
```

---
Q: As I looked closer into the code, it seems like my validation approach duplicates what Setting.Property does (in particular InternalIndex and Deprecated). Is AffixSetting suitable to implement analyzers validation? If not, maybe GroupSettings could be reimplemented as a composition of Settings and verified recursively - so that they relay on the logic behind Properties.

mariha committed:
Disallow users from configuring analyzer versions

mariha committed:
If analyzer with unsupported version is created, warn about version change.

elasticmachine commented:
Pinging @elastic/es-search

jimczi reviewed (commented):
Thanks @mariha! As you noticed already we don't validate analysis setting at all and use an affix setting to retrieve the group settings of each type (`analyzer`, `filter`, ...). Adding proper validation to this group is not a low hanging fruit and will require to change all factories so I wonder if we could do this change differently. Instead of validating at the setting level we can perform the check in the constructor of `AbstractTokenFilterFactory`, `AbstractTokenizerFactory`, ...
Something like:
````
public abstract class AbstractTokenizerFactory extends AbstractIndexComponent implements TokenizerFactory {
    private final String name;

    public AbstractTokenizerFactory(IndexSettings indexSettings, Settings settings, String name) {
        super(indexSettings);
        if (settings.get("version") != null) {
            if (indexSettings.getIndexVersionCreated().after(Version.V_8_0_0)) {
                throw new IllegalArgumentException("[version] is not supported...");
            } else {
                deprecationLogger.deprecatedAndMaybeLog("analysis_version",
                    "[version] is deprecated and will be removed in the next major version");
            }
        }
        this.name = name;
    }

    @Override
    public String name() {
        return name;
    }
}
````
So we'll continue to accept version on indices created before 8.0.0 (master) to ensure backward compatibility and throw an error on new indices that try to set up a version in the analyzer. This will work on closed indices too (the only way to update analysis setting) since we open the index in a dry run to ensure that the new settings will not prevent the index to be opened later.
Does this makes sense ?

mariha committed:
Merge branch 'master' into analyzers-version-settings

mariha committed:
Deprecate index analyzers version. Since v8.0.0 remove it from settings.

Continue to accept version on indices created before 8.0.0 to ensure backward compatibility and throw an error on new indices that try to set up a version in the analyzer. This will work on closed indices too (the only way to update analysis setting) since the index is opened in a dry run to ensure that the new settings will not prevent it to be opened later.

mariha committed:
Cleanup

mariha commented on server/src/main/java/org/elasticsearch/index/analysis/AbstractIndexAnalyzerProvider.java:
```diff
@@ -40,7 +40,10 @@
+
+        Analysis.deprecateSetting("version", Version.V_8_0_0, indexSettings, settings, name, deprecationLogger);
+        // resolve the analysis version based on the version the index was created with
+        this.version = indexSettings.getIndexVersionCreated().luceneVersion;
```
I left version for analyzers because it is passed to lucene and I wasn't sure if it is safe to remove it.

jimczi replied:
++, I don't think we have an official analyzer that uses this information but it might be used in plugins so let's keep it during the deprecation period.

mariha commented:
Thanks for the review @jimczi! I made some corrections based on your feedback, please let me know what do you think.

jimczi commented on server/src/main/java/org/elasticsearch/index/analysis/Analysis.java:
```diff
@@ -78,19 +78,20 @@
+        // check for the setting on the specific index component
+        // and on the index itself as default for all analysis components
+        if (settings.hasValue(settingName) || indexSettings.getSettings().hasValue("index.analysis." + settingName)) {
+            if (indexSettings.getIndexVersionCreated().before(version)) {
```
Where does this logic comes from ? Is it something that you added for this pr ? We should only check that the version is parseable, the validation against the index created version looks like a new constraint that would break settings of users that set the wrong version (accepted today).

mariha replied:
hmm... from the code in your comment? :) I changed it to 
```
if (Version.CURRENT.before(version)) 
    // ...
```

jimczi reviewed (commented):
Thanks @mariha . The pr looks good, I left some comments regarding the deprecation logic.

mariha committed:
log 'version' deprecated message or throw an exception based on current ES version, not version at which the index was created

mariha committed:
pre v8.0.0 parse analysis version as before

mariha commented:
I changed the decision of 'version' being allowed but deprecated or removed and disallowed to be based on ES version. And restored index version parsing logic for older (pre v8.0.0) ES versions. Hope right now it's better @jimczi?

jimczi commented:
Sorry I never came back on this PR. The code has changed in the meantime so I am going to close this PR. Don't hesitate to reopen if you'd like to finish the work (we'll be more reactive this time).
